peps/pep-0746.rst

```diff
@@ -0,0 +1,149 @@
+PEP: 746
+Title: Type checking Annotated metadata
+Author: Adrian Garcia Badaracco <adrian@adriangb.com>
+Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
+Discussions-To: https://discuss.python.org/t/pep-746-typedmetadata-for-type-checking-of-pep-593-annotated/53834
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Created: 20-May-2024
+Python-Version: 3.14
+Post-History: 20-May-2024
+
+Abstract
+========
+
+This PEP proposes a mechanism for type checking metadata that uses
+the :py:data:`typing.Annotated` type. Metadata objects that implement
+the new ``__supports_type__`` protocol will be type checked by static
+type checkers to ensure that the metadata is valid for the given type.
+
+Motivation
+==========
+
+:pep:`593` introduced ``Annotated`` as a way to attach runtime metadata to types.
+In general, the metadata is not meant for static type checkers, but even so,
+it is often useful to be able to check that the metadata makes sense for the given
+type.
+
+Take the first example in :pep:`593`, which uses ``Annotated`` to attach
+serialization information to a field::
+
+   class Student(struct2.Packed):
+       name: Annotated[str, struct2.ctype("<10s")]
+
+Here, the ``struct2.ctype("<10s")`` metadata is meant to be used by a serialization
+library to serialize the field. Such libraries can only serialize a subset of types:
+it would not make sense to write, for example, ``Annotated[list[str], struct2.ctype("<10s")]``.
+Yet the type system provides no way to enforce this. The metadata are completely
+ignored by type checkers.
+
+This use case comes up in libraries like :pypi:`pydantic`, which use
+``Annotated`` to attach validation and conversion information to fields.
+
+Specification
+=============
+
+This PEP introduces a new ``__supports_type__`` protocol that both static and
+runtime type checkers can use to understand if a metadata object in
+``Annotated`` is valid for the given type. Objects that implement this protocol
+must have a method named ``__supports_type__`` that takes a single positional argument and
+returns ``bool``::
+
+    class Int64:
+        def __supports_type__(self, obj: int) -> bool:
+            return isinstance(obj, int)
+
+When a static type checker encounters a type expression of the form ``Annotated[T, M1, M2, ...]``,
+it should enforce that for each metadata element in ``M1, M2, ...``, one of the following is true:
+
+* The metadata element evaluates to a type that does not have a ``__supports_type__`` method; or
+* The metadata element evaluates to an object ``M`` that has a ``__supports_type__`` method, and
+  a call to ``M.__supports_type__(T)`` type checks without errors (i.e., ``T`` is assignable to the
+  parameter type of the ``__supports_type__`` method), and that call does not
+  evaluate to ``Literal[False]``.
```

Comment:
What happens if `__supports_type__` is defined, but left untyped (in particular the return type is omitted)? As pyright tries to infer untyped bodies:

```python
from typing import Any

class Meta:
    def __supports_type__(self, obj: str):
        return False

reveal_type(Meta().__supports_type__(""))  # Type of "Meta().__supports_type__("")" is "Literal[False]"

class Model:
    # Should this be flagged as an error, as `__supports_type__` returns `Literal[False]`?
    foo: Annotated[str, Meta("some_string")]
```

Probably it should be stated that a defined `__supports_type__` but untyped should be ignored?
---

On a side note, I think the following sentence:

> (i.e., ``T`` is assignable to the parameter type of the ``__supports_type__`` method)

is slightly incorrect, as `T` is a type, not a variable, so a type can't be assigned to a parameter type.

(I might be wrong here, please correct me if necessary).

Reply:
It's up to what type checkers infer, similar to any untyped function. However, I'd be OK with omitting the `Literal[False]` special case for simplicity; @adriangb do you think it's important?

I think this usage of "assignable to" is correct under the definitions in python/typing#1743 that we'll likely adopt soon.

Reply:
I think the untyped function is a good case to bring up but in practice I think it won't matter since the implementers will be mostly libraries which I would expect will either fully type their metadata or consider the behavior a bug on their end.

Reply:
Yes, doesn't make really much sense to not type such a method meant for type checking indeed. So I guess the fallback to what type checkers infer is fine!